pkg/splunk/enterprise/monitoringconsole.go

```diff
@@ -0,0 +1,340 @@
+	labels := make(map[string]string)
+	for k, v := range selectLabels {
+		labels[k] = v
+	}
+
+	//create deployment configuration
+	deploymentMC := &appsv1.Deployment{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Deployment",
+			APIVersion: "apps/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      GetSplunkMonitoringConsoleDeploymentName(instanceType, cr.GetNamespace()),
+			Namespace: cr.GetNamespace(),
+		},
+		Spec: appsv1.DeploymentSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: selectLabels,
+			},
+			Replicas: &replicas,
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels:      labels,
+					Annotations: annotations,
+				},
+				Spec: corev1.PodSpec{
+					Affinity:      affinity,
+					Tolerations:   spec.Tolerations,
+					SchedulerName: spec.SchedulerName,
+					Containers: []corev1.Container{
+						{
+							Image:           spec.Image,
+							ImagePullPolicy: corev1.PullPolicy(spec.ImagePullPolicy),
+							Name:            "splunk",
+							Ports:           ports,
+							EnvFrom: []corev1.EnvFromSource{
+								{
+									ConfigMapRef: &corev1.ConfigMapEnvSource{
+										LocalObjectReference: corev1.LocalObjectReference{
+											Name: configMap, //monitoring console env variables configMap
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	if configMapHash == "" {
+		return nil, nil
+	}
+
+	env := []corev1.EnvVar{
+		{Name: "SPLUNK_CONFIGMAP_HASH",
+			Value: configMapHash,
+		},
+	}
+	// append labels and annotations from parent
+	splcommon.AppendParentMeta(deploymentMC.Spec.Template.GetObjectMeta(), cr.GetObjectMeta())
+
+	// update statefulset's pod template with common splunk pod config
+	updateSplunkPodTemplateWithConfig(&deploymentMC.Spec.Template, cr, spec, instanceType, env, secretName)
+
+	return deploymentMC, nil
+}
+
+//ApplyMonitoringConsoleEnvConfigMap creates or updates a Kubernetes ConfigMap for extra env for monitoring console pod
+func ApplyMonitoringConsoleEnvConfigMap(client splcommon.ControllerClient, namespace string, crName string, newURLs []corev1.EnvVar, addNewURLs bool) (*corev1.ConfigMap, error) {
+
+	var current corev1.ConfigMap
+	current.Data = make(map[string]string)
+
+	namespacedName := types.NamespacedName{Namespace: namespace, Name: GetSplunkMonitoringconsoleConfigMapName(namespace, SplunkMonitoringConsole)}
+	err := client.Get(context.TODO(), namespacedName, &current)
+
+	if err == nil {
+		revised := current.DeepCopy()
+		for _, url := range newURLs {
+			_, ok := revised.Data[url.Name]
+			//if new instances are coming up
+			if addNewURLs {
+				//key doesn't exist in the configMap, then add it
+				if !ok {
+					//if new url doesn't exist add it in configmap
+					revised.Data[url.Name] = url.Value
+				} else {
+					//TODO: For "SEARCH_HEAD_URL" need to find a better approach to update configMap, any other data structure?
+					if url.Name == "SPLUNK_SEARCH_HEAD_URL" {
```

Comment:
Sure, During the scaling down/up of shc the "cr.ObjectMeta.DeletionTimestamp" is nill, statefulset manager checks the difference between desiredReplicas and readyReplicas and accordingly updates the stateful set. For MC to update its configMap we check it through deletion timestamp if the deletion is requested or a new resource is getting added. So in case of scaling since the deletion timestamp is always nil, MC doesn't know if we are adding instances or deleting, so first, it deletes all the URLs of that statefulset present in the configMap and then add what are the current URLs present in that statefulset as new. That's why I have separately written the case for Search head here. But I have a TODO here to improvise this logic.